doc/python_reference.rst

```diff
@@ -904,8 +904,17 @@ Simulation
    simulate_stc
    simulate_sparse_stc
    select_source_in_label
+   stc_cosine_score
+   stc_region_localization_error
+   stc_f1_score
+   stc_precision_score
+   stc_recall_score
+   stc_roc_auc_score
+   stc_spacial_deviation
+   stc_peak_position_error
```

Comment:
These all seem to have nearly the same interface, except that only some use `per_sample` and/or `threshold`. Maybe they should all be unified under:
```
stc_score(stc_true, stc_est, method='peak_position_error', per_sample=True, threshold=90)
```
Then there isn't a namespace explosion. Internally we can still use private functions `_stc_cosine_score` to do the calculations, but from an API perspective it's a bit nicer.

Also we generally use numbers instead of strings for percentages, so `90` or `90.` instead of `'90%'`, as it's generally easier to use.

Reply:
Under this API we could even allow `method` to be a single string (return a single score) or a list of string (return a list of scores), or `'all'` to return all scores. But this can be a follow-up PR.

Reply:
> Also we generally use numbers instead of strings for percentages,

Ahh I see now `'90%'` is meant to specify a relative rather than absolute threshold. Personally I would do this with `thresh_kind='absolute' | 'relative'` but I can live with the `str`, so never mind this bit.

I am more concerned with the function explosion here, so maybe we should just discuss just this point. In MNE we usually have a `mode` arg for this sort of thing, even if not all params are used by all modes, and in the docstring you just say which modes they are used for. It prevents namespace explosion. And it's not uncommon in Python -- a rough estimate via `git grep -iI "only used" | wc -l` has for scikit-learn 85 instances, SciPy 34, and in MNE 71, and it probably under-estimates how common it is.

By way of example, to apply the "only use relevant arguments" instead of a single `raw.filter` method we'd instead (at least):
```
raw.filter_lowpass_fir(40)      # raw.filter(None, 40)
raw.filter_lowpass_iir(40)      # raw.filter(None, 40, method='iir')
raw.filter_highpass_fir(1)      # raw.filter(1, None)
raw.filter_highpass_iir(1)      # raw.filter(1, None, method='iir')
raw.filter_bandpass_fir(1, 40)  # raw.filter(1, 40)
raw.filter_bandpass_iir(1, 40)  # raw.filter(1, 40, method='iir')
```
Thinking about parametrizing a script for how you do the filtering:
```
l_freq, h_freq, method = None, 40, 'fir'
```
With the current method the code later is just:
```
raw.filter(l_freq, h_freq, method=method)
```
With the namespace method, you need six conditional branches.

So these in my mind are some advantges of the `mode` way. The disadvantage I see is that the docstring gets more complicated describing the options, and the user needs to understand what's used when. But in the case of this functionality, it all seems very similar and easy to explain when each param is actually used. And you can always make a table in `Notes` to clarify if necessary (but is probably overkill here).

Reply:
@larsoner we need rediscuss this. Let's have all the content first with adequate references to papers, examples that show convincing results etc. We won't have this in 0.18 anyway we have 6 months to iterate/improve. One pro of this approach is that it's easier to discover the different metrics and each function can point to the relevant papers that used this corresponding metric.

btw sklearn has both mechanism. You have all metrics available as public function but you can get them by strings using the scoring/score/make_scorer mechanism.
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html

Reply:
In this sense it could be like the scipy.signal.windows namespace and scipy.signal.get_window, which works okay. I think if we move all of these to their own namespace it could be good, e.g.:

mne.simulation.metrics.cosine_score

I would remove the stc_ as we can say in the submodule doc (and it's obvious from input arguments) that they all operate on STCs.

Reply:
(And if people want a convenience function, we add mme.simulation.compute_metric or something. But maybe we don't need it.)